tests/cases/compiler/inKeywordNarrowingWithNoUncheckedIndexedAccess.ts

```diff
@@ -0,0 +1,37 @@
+// @strict: true
+// @noEmit: true
+// @noUncheckedIndexedAccess: true
+
+declare function invariant(condition: boolean): asserts condition;
+
+function f1(obj: Record<string, string>) {
+    invariant("test" in obj);
+    return obj.test;  // string
+}
+
+function f2(obj: Record<string, string>) {
+    if ("test" in obj) {
+        return obj.test;  // string
+    }
+    return "default";
+}
+
+function f3(obj: Record<string, string>) {
+    obj.test;  // string | undefined
+    if ("test" in obj) {
+        obj.test;  // string
+    }
+    else {
+        obj.test;  // undefined
+    }
+}
+
+function f4(obj: Record<string, string>) {
+    obj.test;  // string | undefined
+    if (obj.hasOwnProperty("test")) {
+        obj.test;  // string
+    }
+    else {
+        obj.test;  // undefined
+    }
+}

```

Comment:
I've hoped that this solution would make it work (somehow through CFA) for at least the `in` variant:
```ts
function f5(obj: Record<string, string>, prop: string) {
    if (obj[prop]) {
        obj[prop]; // actual: string | undefined; expected: string
    }
    if (prop in obj) {
        obj[prop]; // actual: string | undefined; expected: string
    }
}
```

Is this kind of analysis ever supported?

Reply:
`obj[prop]` where `prop` isn’t a hardcoded literal can’t be narrowed in any way to my knowledge.  There are many, many issues attesting to that if you go looking, and it’s considered to be by design.

Reply:
Everything gets duped to #10530 which is kind of misleading since that issue was originally about bracketed access with a literal key (which has since been fixed), but the picture that develops after enough hunting is that it’s a performance sinkhole.  See e.g. https://github.com/microsoft/TypeScript/issues/25109#issuecomment-398926483

Reply:
@Andarist It works when `prop` is declared as a `const` with a singleton literal type, e.g. `const prop = "foo"`, but no support beyond that. We could conceivably support non-singleton-literal-type const variables, but it isn't trivial.

Reply:
Huh, I wasn’t even aware _that_ worked.

Reply:
> There are many, many issues attesting to that if you go looking, and it’s considered to be by design.

I understand that this seems difficult to implement and we might never get this feature, but I'm having a hard time understanding how this could be "by design". What I mean is, this seems like fairly straight forward JS and as far as I understand TypeScript's design goals, we really should be able to support (type correctly) this kind of code. I really am sympathetic to the fact that it might not be feasible to do in the current implementation, I'm just challenging the idea that we _want_ it to be this way.

Reply:
“By design” = ”intentional decision made during the design process”.  Some such decisions are known compromises.  I don’t use the phrase to imply it’s what the *user* wants, because nobody can even define that except for the individual users themselves, and the team can’t go around asking every single user before they’re allowed to label things “working as intended”. :wink:

Reply:
Sure there always needs to be compromises made in the design of the language, but this isn't that. This seems to be a limitation of the current implementation (an implementation detail if you will), not some inherent limitation in the design of the language itself.

Reply:
I mean, the implementation is still part of the design *of the TypeScript compiler* in particular.  Obviously a *language* wouldn't be intentionally designed that way in isolation, but that's a moot point since there's no TypeScript language specification.  There used to be a formal spec a long time ago, but nowadays the implementation is the specification, and that's an entirely pragmatic design process.  If the implementation is designed a certain way, in practice that means the language is too.

Reply:
...at least until someone gets fed up with the limitations and makes an alternative implementation. 😅